### ISSUE TITLE
Make circuits occur in coop MM

### ISSUE THREAD

**Issue by GoogleFrog** [ZeroK-RTS/Zero-K-Infrastructure#2498]
Lobbies should send infra a map of difficulties -> aiNames. Server then inserts the relevant ones into the startscript with the appropriate owners. For now, don't agonise over questions such as who exactly hosts the circuits.

**Comment by GoogleFrog** [comment]
The lobby would send a table such as this in an appropriate MM command, or perhaps even on connecting to the server `{novice : "circuitAINoviceStable32", hard : "circuitAiHardStable32"}`.

With this information the server can know what to put in the startscripts.

**Comment by DeinFreund** [comment]
Sounds good, let me know how the message is called and laid out.

**Comment by Licho1** [comment]
No, do not implement this. Lobby cannot specify difficulties. Make it stored on server side or in somet dynamic config string variable.
If you do it this way its hackable and people could inject failing/crashing/custom AIs

**Comment by DeinFreund** [comment]
As long as AIs are run on the client, they can always be tampered with.

**Comment by Licho1** [comment]
But in this case one person would infect the other - its coop. This data should be on server or part of the game itself. 

If its part of game itself it means translation happens somewhere in chobbies.

In that case the server would only send:
AI:easy
AI:normal
AI:hard

and chobby will translate to corresponding AI based on current state.

**Comment by Anarchid** [comment]
>  If you do it this way its hackable and people could inject failing/crashing/custom AIs

I don't see the incentive.

**Comment by sprunk** [comment]
shaman would eventually do it for the lulz

**Comment by Anarchid** [comment]
Converting shaman into an AI dev sounds good to me!

**Comment by Licho1** [comment]
Think about it the other way, there is perhaps no reason for server to know actual AI name. It could refer to it using well known and agreed alias. That way each version of ZK could have slightly different mapping, or there could be more configs for given AI etc. 
Server will still know which general class of AI was used (medium, hard etc) and could learn it's skill using emo/whr

**Comment by sprunk** [comment]
>In that case the server would only send:
>AI:easy
>AI:normal
>AI:hard
>and chobby will translate to corresponding AI based on current state.

Effectively this would mean that AI files have to be called "AI_hard" etc. on all player local machines, because server sends AI file paths in startscript and Spring just opens up `./AI/<name from startscript>.dll`.

**Comment by GoogleFrog** [comment]
@Licho1 how is the server going to generate startscripts if is not told the AI names that match the clients Spring VFS?

**Comment by GoogleFrog** [comment]
> No, do not implement this. Lobby cannot specify difficulties. Make it stored on server side or in somet dynamic config string variable.

The lobby is not specifying difficulties. The lobby is telling the server the names of its AIs and which difficulty they correspond to. Any locally hosted AI is hackable, so I'd ignore this issue (or make admin rules if the AIs are used for planetwars).

**Comment by GoogleFrog** [comment]
Perhaps the wrapper could include a way to hash the AI files.

**Comment by Licho1** [comment]
Right because its native AI we need to know external AI name. 
I don't think lobbies can send AIs, when would they do it? On login? For what game version? It sounds complicated.
Let's have a list of AIs, possibly keyed by Zero-K version in some server database?

**Comment by Anarchid** [comment]
> Perhaps the wrapper could include a way to hash the AI files.
The wrapper is also hackable.

**Comment by Licho1** [comment]
Wrapper way you are suggesting is also possible.
In that case wrapper would for example access the game file ,look for lua table with mapping "universal anme" -> "concrete name" and apply that.

**Comment by GoogleFrog** [comment]
Matchmaker games should always use the server default game and engine versions, so the version issue should be avoidable. Lobbies that do not have the correct game or engine version already implement checks that prevent them from joining MM games while the mismatch persists.

Lobbies can send AIs, doing it on login would work. We could also send them closer to the time of the MM game in whatever command is used to tell the server that the player has accepted the game. I think it is better for the lobby to send it to the server, rather than the wrapper talk to the server, because the lobby already keeps track of which AIs can be used on which engine versions.

**Comment by Licho1** [comment]
Ok. So what exactly do you propose? In what moment will lobbies send data to server. Will server need to track this information for each user connected?

**Comment by GoogleFrog** [comment]
Add a table to AreYouReadyResponse. The table can have the keys beginner, novice, easy, normal, hard, and brutal. The lobby will put strings in the table corresponding to the appropriate AI for each difficulty level. Keys may not exists (perhaps that AI is missing). The table in AreYouReadyResponse can be used to generate coop MM games that may be created from the response.

This avoids tracking the information for every user and solves cases where the AIs may change.

**Comment by Licho1** [comment]
This is not sufficient. I can imagine scenarios where mm-like battle with AI starts without this input.
For example in planetwars. 
It has to be sent on startup but then for what game version? 
Maybe it will be best if server just keeps this - for each ZK version table of AIs available + their rank (easy, hard, normal).
This information can either be hand edited on website or extracted from some game file (technically more complex).

We always ship same AIs to all users and they only change with ZK version change

**Comment by Anarchid** [comment]
> I can imagine scenarios where mm-like battle with AI starts without this input.
For example in planetwars.

Why would a PW AI battle be started without this table?

> Maybe it will be best if server just keeps this - for each ZK version table of AIs available + their rank (easy, hard, normal).

Maybe there's not even a need for to support separate lists for "each ZK".

There's just one ZK+Chobby+Wrapper+Engine combination that is valid for MM at any given time.  It's not possible to enter MM without having this combination. This means that only one set of AI's has to be supported at any given time, and this could be a server config (if this changes with required engine).

**Comment by GoogleFrog** [comment]
I still hear rumors of people running different lobbies, so I want to make sure that the people assigned to run each AI actually have the AI.

The planetwars protocol has clients send information to the server at some point to indicate having signed up for planetwars. The table can be inserted here.

**Comment by Licho1** [comment]
Sounds way more complicated than needed to be. Instead we could simply ask for lobby version on login (unless we already do) and then just ignore invalid version people..
Or just screw people who have random lobby .. i mean how can they even join MM if they have invalid lobby?

**Comment by GoogleFrog** [comment]
I think we do ask for lobby version. Are you suggesting some sort of dynamic config for the circuit strings? An issue is that the string changes depending on OS and 32/64 bit.

**Comment by Licho1** [comment]
Well I don't see much better option.
We could send it on chobby  startup, but then its unclear for what game version it is. We could send it on "ready" but i can imagine situations where it does not ask, we would be unflexible.

**Comment by GoogleFrog** [comment]
I think sending it on chobby startup is the way to go. The AIs don't care about game version, and you can't join MM if you engine version doesn't match the server.

**Comment by Licho1** [comment]
I am thinking about scenario where game version changes mid game and we update circuit for new version - to new circuit which is not in old version ..

**Comment by GoogleFrog** [comment]
I made a lagmonitor for circuitAIs that shares units around if a player drops (deleting an AI). Also, add a CAI to every MM and planetwars AI game with custom keys nocommander=1 and backupai=1 to create an AI that just steps in to take over from the circuits in the event of all circuit hosts dropping. The circuit hosts should be spread evenly between the involved players to avoid this as much as possible.
